Rubberduck.Parsing/Symbols/AccessibilityCheck.cs

```diff
@@ -4,94 +4,146 @@ public static class AccessibilityCheck
-                       || moduleOrMember.Accessibility == Accessibility.Public
-                       || moduleOrMember.Accessibility == Accessibility.Friend
-                       || moduleOrMember.Accessibility == Accessibility.Implicit);
-        }
+            private static bool IsTheSameModule(Declaration callingModule, Declaration calleeModule)
+            {
+                return calleeModule.Equals(callingModule);
+            }
+
+            private static bool IsEnclosingProject(Declaration callingProject, Declaration calleeModule)
+            {
+                return calleeModule.ParentScopeDeclaration.Equals(callingProject);
+            }
+
+
 
         public static bool IsMemberAccessible(Declaration callingProject, Declaration callingModule, Declaration callingParent, Declaration calleeMember)
         {
-            if (IsEnclosingModule(callingModule, calleeMember))
+            if (calleeMember == null)
+            {
+                return false;
+            }    
+            else if (IsEnclosingModuleOfInstanceMember(callingModule, calleeMember))
             {
                 return true;
             }
-            var callerIsSubroutineOrProperty = callingParent.DeclarationType.HasFlag(DeclarationType.Property)
-                || callingParent.DeclarationType.HasFlag(DeclarationType.Function)
-                || callingParent.DeclarationType.HasFlag(DeclarationType.Procedure);
-            var calleeHasSameParent = callingParent.Equals(callingParent.ParentScopeDeclaration);
-            if (callerIsSubroutineOrProperty && calleeHasSameParent)
+            else if (IsLocalMemberOfTheCallingSubroutineOrProperty(callingParent, calleeMember))
             {
-                return calleeHasSameParent;
+                return true;
             }
             var memberModule = Declaration.GetModuleParent(calleeMember);
             if (IsModuleAccessible(callingProject, callingModule, memberModule))
             {
-                if (calleeMember.DeclarationType.HasFlag(DeclarationType.EnumerationMember) || calleeMember.DeclarationType.HasFlag(DeclarationType.UserDefinedTypeMember))
+                if (calleeMember.DeclarationType.HasFlag(DeclarationType.EnumerationMember) || calleeMember.DeclarationType.HasFlag(DeclarationType.UserDefinedTypeMember)) 
+                {
+                    return true;
+                }
+                else if (IsEnclosingProject(callingProject, memberModule) && IsAccessibleThroughoutTheSameProject(calleeMember))
                 {
-                    return IsValidAccessibility(calleeMember.ParentDeclaration);
+                    return true;
                 }
                 else
                 {
-                    return IsValidAccessibility(calleeMember);
+                    return HasPublicScope(calleeMember);
                 }
             }
-            return false;
+            else
+            {
+                return false;
+            }
```

Comment:
This whole chain could be compressed into one boolean expression. Extracting subexpressions into methods is probably applicable again